cadence/nft-lock-box/NFTLockBox.cdc

```diff
@@ -0,0 +1,168 @@
+import NonFungibleToken from {{{ imports.NonFungibleToken }}}
+import FungibleToken from {{{ imports.FungibleToken }}}
+
+pub contract NFTLockBox {
+
+    // The Claimed event is emitted when an NFT
+    // is successfully claimed from a lock box.
+    pub event Claimed(
+        nftType: Type,
+        nftID: UInt64,
+        from: Address?,
+        recipient: Address
+    )
+
+    // These are the default storage and public paths
+    // defined by the lock box contract.
+    //
+    // However, a lock box owner may store multiple lock boxes 
+    // in the same account using non-default storage paths.
+    pub let DefaultLockBoxStoragePath: StoragePath
+    pub let DefaultLockBoxPublicPath: PublicPath
+
+    // This is the publicly-available interface
+    // that allows users to claim NFTs from a lock box.
+    pub resource interface LockBoxPublic {
+        pub fun claim(
+            id: UInt64, 
+            address: Address,
+            signature: [UInt8]
+        )
+    }
+
+    // A lock box is an NFT collection that maintains
+    // a one-to-one mapping between NFTs and ECDSA public keys.
+    //
+    // A user can withdraw a given NFT if they have its corresponding
+    // private key, referred to as the "claim key".
+    //
+    // The lock box owner can distribute claim keys (e.g. via email or QR code)
+    // in order to facilitate an airdrop without requiring users to
+    // have an existing Flow account address.
+    pub resource LockBox: LockBoxPublic {
+
+        // A capability to the underlying base NFT collection
+        // that will store the claimable NFTs.
+        access(self) let collection: Capability<&{NonFungibleToken.Provider, NonFungibleToken.Receiver}>
+        
+        // When moving a claimed NFT in an account,
+        // the lock box will deposit the NFT into 
+        // the NonFungibleToken.CollectionPublic linked at this public path.
+        access(self) let receiverPath: PublicPath
```

Comment:
Do we need to store this? can't we take this from the metadata views automatically?

Reply:
We don't need to. But I'm hesitant to make the lockbox have an implicit dependency on the `NFTCollectionData` view.

What about something like this:

```cadence
let collectionData = resolveCollectionDataView() // either get the collection data view or construct it on the fly

let lockBox = NFTLockBox.createLockBox(collection: collection, collectionData: collectionData)

// User can then use helpers like:

let collection = lockBox.createEmptyCollection()
```

In the majority of cases, the `collectionData` parameter could be automatically resolved as a view. But for projects that don't support it, there's still a way to use the lock box contract.

Reply:
I started implementing the above, but once I got into, making `NFTLockBox` dependent on `MetadataViews` felt more like a code smell to me.

I personally prefer the loose coupling of just `receiverPath`. We can always resolve the path from `NFTCollectionData` inside a transaction. What do you think?